AtReconstruction/AtFitter/OpenKF/kalman_filter/TrackFitterUKF.h

```diff
@@ -0,0 +1,375 @@
+   }
+
+   ///
+   /// @brief set the measurement noise covariance R to be used in the update step
+   ///
+   void setCovarianceR(const Matrix<DIM_N, DIM_N> &matR)
+   {
+      m_matR = matR; // Store the measurement noise covariance matrix
+   }
+
+   /// Add state vector (m_vecX) to the augment state vector (m_vecXa) and also
+   /// add covariance matrix (m_matP) to the augment covariance (m_matPa).
+   void updateAugWithState()
+   {
+      // Copy state vector to augmented state vector
+      for (int32_t i{0}; i < DIM_X; ++i) {
+         m_vecXa[i] = m_vecX[i];
+      }
+
+      // Copy state covariance matrix to augmented covariance matrix
+      for (int32_t i{0}; i < DIM_X; ++i) {
+         for (int32_t j{0}; j < DIM_X; ++j) {
+            m_matPa(i, j) = m_matP(i, j);
+         }
+      }
+   }
+
+   std::array<float32_t, DIM_V> calculateProcessNoiseMean()
+   {
+      // Calculate the expectation value of the process noise using the current value of the state vector m_vecX
+      std::array<float32_t, DIM_V> processNoiseMean{0};
+
+      // TODO: Set the mean energy loss based on the momentum and particle type. Probably best to track stopping power?
+      return processNoiseMean;
+   }
+
+   Matrix<DIM_V, DIM_V> calculateProcessNoiseCovariance()
+   {
+      // Calculate the process noise covariance matrix
+      Matrix<DIM_V, DIM_V> matQ{Matrix<DIM_V, DIM_V>::Zero()};
+      matQ = m_matQ; // Use the stored process noise covariance matrix
+
+      // TODO: Set the process noise covariance for angular straggle and energy loss.
+      return matQ;
+   }
+
+   void updateAugWithProcessNoise()
+   {
+      auto processNoiseMean = calculateProcessNoiseMean();
+      m_matQ = calculateProcessNoiseCovariance();
+
+      // Add the mean process noise to the augmented state vector
+      for (int32_t i{0}; i < DIM_V; ++i) {
+         m_vecXa[DIM_X + i] = processNoiseMean[i];
+      }
+
+      // Add process noise covariance to the augmented covariance matrix
+      const int32_t S_IDX{DIM_X};
+      const int32_t L_IDX{S_IDX + DIM_V};
+
+      for (int32_t i{S_IDX}; i < L_IDX; ++i) {
+         for (int32_t j{S_IDX}; j < L_IDX; ++j) {
+            m_matPa(i, j) = m_matQ(i - S_IDX, j - S_IDX);
+         }
+      }
+   }
+
+   ///
+   /// @brief update the augmented state vector and covariance matrix
+   /// This functions fully updates the augmented state vector (m_vecXa) and covariance matrix (m_matPa)
+   /// by setting both the state vector and process noise components.
+   ///
+   void updateAugmentedStateAndCovariance()
+   {
+      updateAugWithState();
+      updateAugWithProcessNoise();
+   }
+
+   ///
+   /// @brief state prediction step of the unscented Kalman filter (UKF).
+   /// @param predictionModelFunc callback to the prediction/process model
+   /// function
+   ///
+   template <typename PredictionModelCallback>
+   void predictUKF(PredictionModelCallback predictionModelFunc)
```

Comment:
This takes in the prediction function. It will be defined in the task, for now there is a test class that is compiled and run with GTest for the initial implementation